File: ACM/LanQiao/2940/2940.cpp
/*
 * @Author: NEFU AB-IN
 * @Date: 2023-06-09 18:08:14
 * @FilePath: \LanQiao\2940\2940.cpp
 * @LastEditTime: 2023-06-09 18:08:20
 */
#include <cstring>
#include <iostream>
#include <math.h>
using namespace std;
typedef long long ll;
const int N = 2e5 + 5, M = 18;

int n, w[N];
int f[N][M];

int gcd(int a, int b)
{
    return b ? gcd(b, a % b) : a;
}

void init()
{
    for (int j = 0; j < M; j++)
        for (int i = 1; i + (1 << j) - 1 <= n; i++)
        {
            if (!j)
                f[i][j] = w[i];
            else
                f[i][j] = gcd(f[i][j - 1], f[i + (1 << j - 1)][j - 1]);
        }
}

int query(int l, int r)
{
    int len = r - l + 1;
    int k = log(len) / log(2);
    return gcd(f[l][k], f[r - (1 << k) + 1][k]);
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cin >> n;
    for (int i = 1; i <= n; i++)
        cin >> w[i];

    init();

    int pre = 0, ans = 0;
    for (int i = 1; i <= n; i++)
    {
        int l = pre + 1, r = i;
        while (l < r)
        {
            int mid = l + r >> 1;
            if (query(mid, i) >= i - mid + 1)
                r = mid;
            else
                l = mid + 1;
        }
        if (query(l, i) == i - l + 1)
        {
            ans++;
            pre = i;
        }
    }
    cout << ans;

    return 0;
}

File: ACM/LanQiao/2940/2940.py
'''
Author: NEFU AB-IN
Date: 2023-06-09 18:00:12
FilePath: \LanQiao\2940\2940.py
LastEditTime: 2023-06-09 20:26:52
'''
# import
from sys import setrecursionlimit, stdin, stdout, exit
from collections import Counter, deque
from heapq import heapify, heappop, heappush, nlargest, nsmallest
from bisect import bisect_left, bisect_right
from datetime import datetime, timedelta
from string import ascii_lowercase, ascii_uppercase
from math import log, gcd, sqrt, fabs, ceil, floor


class sa:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __lt__(self, a):
        return self.y < a.y


# Final
N = int(2e5 + 10)
M = 20
INF = int(2e9)

# Define
setrecursionlimit(INF)
input = lambda: stdin.readline().rstrip("\r\n")  # Remove when Mutiple data
read = lambda: map(int, input().split())
LTN = lambda x: ord(x.upper()) - 65  # A -> 0
NTL = lambda x: ascii_uppercase[x]  # 0 -> A

# —————————————————————Division line ——————————————————————
dp = [[0] * M for _ in range(N)]
Log = [0] * N
a = [0] * N


def init():
    for j in range(M):
        i = 1
        while i + (1 << j) - 1 <= n:
            if j == 0:
                dp[i][j] = a[i]
            else:
                dp[i][j] = gcd(dp[i][j - 1], dp[i + (1 << (j - 1))][j - 1])
            i += 1
    for i in range(2, N):
        Log[i] = Log[i // 2] + 1


def query(l, r):
    k = Log[r - l + 1]
    return gcd(dp[l][k], dp[r - (1 << k) + 1][k])


n, = read()
a[1:] = read()

ans = []
init()

for i in range(1, n + 1):
    l, r = i, n
    while l < r:
        mid = l + r + 1 >> 1
        if query(i, mid) >= mid - i + 1:
            l = mid
        else:
            r = mid - 1
    if query(i, l) == l - i + 1:
        ans.append(sa(i, l))

cnt = 1
if len(ans) == 0:
    print(0)
else:
    ans.sort()
    tmp = ans[0].y
    for i in ans:
        if i.x > tmp:
            cnt += 1
            tmp = i.y
    print(cnt)
